File: 19.MergeSort.py
# Merge Sort is a divide-and-conquer, comparison-based sorting algorithm that repeatedly splits the array into halves,
# sorts each half, and then merges the sorted halves.
# It guarantees O(n log n) time in best, average, and worst cases and is a stable sort (preserves equal-element order).

# Explanation

# Idea: Recursively split the list into halves until each piece has 0 or 1 element (already sorted).
# Then merge those small sorted lists back together in order.

# Divide & Conquer: Depth of recursion ≈ log₂(n). Each level performs O(n) work merging → overall O(n log n).

# Space: Typical implementations allocate extra arrays for merging → O(n) additional space (not in-place).

# Stable: If merging uses <= when comparing, elements equal in value retain their relative order.

# Variants: Top-down (recursive) shown here; bottom-up (iterative) is also possible.

def merge(left, right):
    merged = []
    i = 0
    j = 0

    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1

    if i < len(left):
            merged.extend(left[i:])
    if j < len(right):
            merged.extend(right[j:])

    return merged


def merge_sort(arr):
    if len(arr) <= 1:
        return arr

    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    sorted_left = merge_sort(left_half)
    sorted_right = merge_sort(right_half)

    return merge(sorted_left, sorted_right)


if __name__ == "__main__":
    arr = [38, 27, 43, 3, 9, 82, 10]
    print("Original:", arr)
    sorted_arr = merge_sort(arr)
    print("Sorted: ", sorted_arr)

File: 19.MergeSort.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio        */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class MergeSortDemo {

    // Merge two sorted arrays into one sorted array
    public static int[] merge(int[] left, int[] right) {
        int[] merged = new int[left.length + right.length]; // final merged array
        int i = 0;  // pointer for left
        int j = 0;  // pointer for right
        int k = 0;  // pointer for merged

        // compare elements from left and right, copy the smaller into merged
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                merged[k] = left[i];  // take from left
                i++;
            } else {
                merged[k] = right[j]; // take from right
                j++;
            }
            k++;
        }

        // copy remaining elements from left, if any
        while (i < left.length) {
            merged[k] = left[i];
            i++;
            k++;
        }

        // copy remaining elements from right, if any
        while (j < right.length) {
            merged[k] = right[j];
            j++;
            k++;
        }

        return merged; // return merged sorted array
    }

    // Merge Sort function
    public static int[] mergeSort(int[] arr) {
        // base case: if array has 0 or 1 elements, it is already sorted
        if (arr.length <= 1) {
            return arr;
        }

        // find mid index to split array into two halves
        int mid = arr.length / 2;

        // create left and right halves
        int[] left = Arrays.copyOfRange(arr, 0, mid); // elements [0..mid-1]
        int[] right = Arrays.copyOfRange(arr, mid, arr.length); // elements [mid..end]

        // recursively sort both halves
        int[] sortedLeft = mergeSort(left);
        int[] sortedRight = mergeSort(right);

        // merge sorted halves
        return merge(sortedLeft, sortedRight);
    }

    public static void main(String[] args) {
        int[] arr = {38, 27, 43, 3, 9, 82, 10};  // original unsorted array
        System.out.println("Original: " + Arrays.toString(arr)); // print original

        int[] sortedArr = mergeSort(arr); // sort array using merge sort

        System.out.println("Sorted:   " + Arrays.toString(sortedArr)); // print sorted
    }
}
